lib/node_modules/@stdlib/utils/parse-dsv/test/test.js

```diff
@@ -0,0 +1,144 @@
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+'use strict';
+
+// MODULES //
+
+var tape = require( 'tape' );
+var string2buffer = require( '@stdlib/buffer/from-string' );
+var DSVParser = require( './../lib' );
+var fixtures = require( './js/fixtures/fixtures.json' );
+
+tape( 'main export is a function', function test( t ) {
+	t.ok( true, __filename );
+	t.equal( typeof DSVParser, 'function', 'main export is a function' );
+	t.end();
+});
+
+tape( 'the implementation properly parses csv files', function test( t ) {
+	var delimiter;
+	var quotechar;
+	var expected;
+	var headers;
+	var parsed;
+	var input;
+	var data;
+	var pd;
+	var i;
+
+	for ( i = 0; i < fixtures.length; i++ ) {
+		data = fixtures[ i ].data;
+		delimiter = data.delimiter;
+		quotechar = data.quotechar;
+		input = data.input;
+		expected = fixtures[ i ].expected;
+		pd = new DSVParser( delimiter, 'utf8', quotechar );
+
+		// transform solutions with headers to 2d array versions
+		if ( !( expected[ 0 ] instanceof Object ) ) {
+			expected = [ expected ];
+		} else if ( !Array.isArray( expected[ 0 ] ) ) {
+			headers = Object.keys( expected[ 0 ] );
+			expected = expected.map( function map( e ) {
+				return Object.values( e );
+			});
+			expected.unshift( headers );
+		}
+		try {
+			parsed = pd.incrparse(string2buffer(input));
+			parsed = pd.incrparse(null);
+
+			t.deepEqual( parsed, expected, 'should have same output when parsing ' + input);
+		} catch ( e ) {
+			t.ok( false, 'Recieved error: ' + e + ' when parsing ' + input + ' with delimiter ' + delimiter);
+		}
+	}
+	t.end();
+});
+
+tape( 'the implementation properly emits rows and end of streams', function test( t ) {
+	var delimiter;
+	var quotechar;
+	var expected;
+	var headers;
+	var parsed;
+	var input;
+	var data;
+	var pd;
+
+	(function recursiveHelper( i ) {
+		if ( i === fixtures.length ) {
+			t.end();
+			return;
+		}
+
+		data = fixtures[ i ].data;
+		delimiter = data.delimiter;
+		quotechar = data.quotechar;
+		input = data.input;
+		expected = fixtures[ i ].expected;
+		pd = new DSVParser( delimiter, 'utf8', quotechar );
+
+		if ( !( expected[ 0 ] instanceof Object ) ) {
+			expected = [ expected ];
+		} else if ( !Array.isArray( expected[ 0 ] ) ) {
+			headers = Object.keys( expected[ 0 ] );
+			expected = expected.map( function map( e ) {
+				return Object.values( e );
+			});
+			expected.unshift( headers );
+		}
+
+		parsed = [];
+		pd.on( 'row', function listener( row, rowCount ) {
+			t.deepEqual( row, expected[ rowCount ], 'emits rows properly' );
+			parsed.push( row );
+		});
+		pd.on( 'end', function callback() {
+			t.deepEqual( parsed, expected, 'emits end of stream properly' );
+			recursiveHelper( i + 1 );
+		});
+
+		pd.incrparse( string2buffer( input ) );
+		pd.incrparse( null );
+	})( 0 );
+});
+
+tape( 'the implementation properly throws errors', function test( t ) {
```

Comment:
May be better to emit, rather than throw, parse errors. This would essentially punt error handling to userland.